### PR TITLE
[OBSDEF-4285 ] Added detail pane for datagrid

### PR DESCRIPTION
**Pull request by akate1**
**Description:**
- Added detail pane feature in datagrid as per https://clarity.design/documentation/datagrid/detail-pane 

**JIRA:**
- https://jira.cec.lab.emc.com/browse/OBSDEF-4285

**Testing:**
detailPane

### CODE REVIEW

File: src/datagrid/DataGrid.tsx [hunk 1]
```diff
@@ -759,9 +841,79 @@ export class DataGrid extends React.PureComponent<DataGridProps, DataGridState>
...
+                    <Button
+                        key={`${id}-${"detail-toggle"}-${rowID}`}
+                        defaultBtn={false}
+                        onClick={() => rowID !== undefined && this.handleDetailPaneToggle(rowID)}
+                        aria-haspopup="dialog"
+                        className={classNames([ClassNames.DATAGRID_DETAIL_CARET_BUTTON, isOpen && ClassNames.IS_OPEN])}
+                        aria-expanded={isOpen}
+                        aria-controls={`clr-id-${rowID}`}
+                        icon={{
+                            shape: "angle-double",
```
Comment (maheshkulkarni4) [inline src/datagrid/DataGrid.tsx:907]: can we create enum with these shapes ?  We have many such shapes I think in our codebase.

File: src/datagrid/DataGrid.tsx [hunk 2]
```diff
@@ -1034,43 +1199,62 @@ export class DataGrid extends React.PureComponent<DataGridProps, DataGridState>
...
-                        `ng-tns-c96-${index}`,
-                        isExpandableRow && ClassNames.DATAGRID_EXPAND_ANIMATION,
-                    ])}
-                >
+                {isExpandableRow ? (
                     <div
-                        className={classNames([ClassNames.DATAGRID_ROW_MASTER, ClassNames.DATAGRID_NG_STAR_INSERTED])}
-                        role="row"
+                        className={classNames([
+                            `ng-tns-c96-${index}`,
```
Comment (maheshkulkarni4) [inline src/datagrid/DataGrid.tsx:1215]: how this c96 calculated ? does it have any significance ?

File: src/datagrid/DataGrid.tsx [hunk 3]
```diff
@@ -1098,6 +1282,85 @@ export class DataGrid extends React.PureComponent<DataGridProps, DataGridState>
         );
     }
 
+    private buildDetailPaneForRow() {
```
Comment (maheshkulkarni4) [inline src/datagrid/DataGrid.tsx:1285]: Add some documentation

File: src/datagrid/DataGrid.tsx [hunk 4]
```diff
@@ -1034,43 +1199,62 @@ export class DataGrid extends React.PureComponent<DataGridProps, DataGridState>
...
+                        {this.buildRowCells(row, isExpandableRow, isRowWithDetailPane)}
                     </div>
+                ) : (
+                    this.buildRowCells(row, isExpandableRow, isRowWithDetailPane)
+                )}
+            </div>
+        );
+    }
+
+    private buildRowCells(row: DataGridRow, isExpandableRow: boolean, isRowWithDetailPane: boolean) {
```
Comment (maheshkulkarni4) [inline src/datagrid/DataGrid.tsx:1228]: Add some documentation

File: src/datagrid/DataGrid.tsx [hunk 5]
```diff
@@ -1098,6 +1282,85 @@ export class DataGrid extends React.PureComponent<DataGridProps, DataGridState>
...
+                                className={classNames([
+                                    ClassNames.DATAGRID_DETAIL_PANE_CONTENT,
+                                    ClassNames.DATAGRID_NG_STAR_INSERTED,
+                                ])}
+                                id={`${"clr-id-"}-${rowID}`}
+                                aria-describedby={`clr-id-${rowID}-title`}
+                                tabIndex={-1}
+                            >
+                                <div className={ClassNames.CLR_SR_ONLY} />
+                                <div _ngcontent-byx-c249="" className={ClassNames.DATAGRID_DETAIL_HEADER}>
```
Comment (maheshkulkarni4) [inline src/datagrid/DataGrid.tsx:1327]: c249 ?

File: src/datagrid/DataGrid.tsx [hunk 6]
```diff
@@ -1098,6 +1282,85 @@ export class DataGrid extends React.PureComponent<DataGridProps, DataGridState>
...
+                                        className={ClassNames.DATAGRID_DETAIL_HEADER_TITLE}
+                                        id={`clr-id-${rowID}-title`}
+                                    >
+                                        {title}
+                                    </div>
+                                    <div className={ClassNames.DATAGRID_DETAIL_PANE_CLOSE}>
+                                        <Button
+                                            aria-label="Close"
+                                            className={ClassNames.PAGINATION_NEXT}
+                                            icon={{shape: "close"}}
```
Comment (maheshkulkarni4) [inline src/datagrid/DataGrid.tsx:1338]: One more shape.

File: src/datagrid/DataGrid.tsx [hunk 7]
```diff
@@ -1098,6 +1282,85 @@ export class DataGrid extends React.PureComponent<DataGridProps, DataGridState>
...
+                                            onClick={() => rowID !== undefined && this.handleDetailPaneToggle(rowID)}
+                                        />
+                                    </div>
+                                </div>
+                                {/* close header*/}
+                                <div ngcontent-byx-c249="" className={ClassNames.DATAGRID_DETAIL_BODY}>
```
Comment (maheshkulkarni4) [inline src/datagrid/DataGrid.tsx:1344]: why is empty content needed > 
 **ngcontent-byx-c249=""**

File: src/datagrid/DataGridStoriesData.tsx [hunk 8]
```diff
@@ -421,6 +485,29 @@ export const getPageDataForSelectedRows = (pageIndex: number, pageSize: number):
         }, 2000);
     });
 };
+
+// Function to get data for page based on pagenumber
+export const getPageDataForDetailPane = (pageIndex: number, pageSize: number): Promise<DataGridRow[]> => {
+    return new Promise((resolve, reject) => {
```
Comment (maheshkulkarni4) [inline src/datagrid/DataGridStoriesData.tsx:491]: There are some usecases, when we dont have data in Left side pane... in that case the details panel should show some message.  Can we introduce a prop to handle such usecase. Think over it

Usecase:  You  launch Details pane with Namespace which has object stores, once details pane is open, you swich namespace which dont have any object stores.  In that case you done have first row to show data in details pane.  

Worth thinking if fix will be in this component or Parent

**Review by maheshkulkarni4 — APPROVED** [review]
Overall the component looks good,  UI is very appeling. 
Add stories for following usecase
e.g. We have 10 rows
1) First row selected
2) We have 6th row selected
3) we have 10th row selected
4) Check following too 
There are some usecases, when we dont have data in Left side pane... in that case the details panel should show some message. Can we introduce a prop to handle such usecase. Think over it

Usecase: You launch Details pane with Namespace which has object stores, once details pane is open, you swich namespace which dont have any object stores. In that case you done have first row to show data in details pane.

Worth thinking if fix will be in this component or Parent

File: src/datagrid/DataGrid.tsx (hunk 2 above)
Comment (akate1) [reply src/datagrid/DataGrid.tsx:1215]: This is CSS class name coming from clarity. c96 is same for all rows not calculated

File: src/datagrid/DataGrid.tsx (hunk 1 above)
Comment (akate1) [reply src/datagrid/DataGrid.tsx:907]: We support quite lot of shapes. Can't add enum for all of them. Plus this seem unnecessary. Except some common icon like close we are using other shapes only one time

File: src/datagrid/DataGridStoriesData.tsx (hunk 8 above)
Comment (akate1) [reply src/datagrid/DataGridStoriesData.tsx:491]: @maheshkulkarni4 With current implementation of detail pane if you switch namespace when detail pane is open then datagrid will get re-rendered and detail pane will get close.

**Comment by akate1** [comment]
> Overall the component looks good, UI is very appeling.
> Add stories for following usecase
> e.g. We have 10 rows
> 
> 1. First row selected
> 2. We have 6th row selected
> 3. we have 10th row selected
> 4. Check following too
>    There are some usecases, when we dont have data in Left side pane... in that case the details panel should show some message. Can we introduce a prop to handle such usecase. Think over it
> 
> Usecase: You launch Details pane with Namespace which has object stores, once details pane is open, you swich namespace which dont have any object stores. In that case you done have first row to show data in details pane.
> 
> Worth thinking if fix will be in this component or Parent

@maheshkulkarni4 I have already added story of row selection and detail pane working together
image

File: src/datagrid/DataGrid.tsx (hunk 5 above)
Comment (akate1) [reply src/datagrid/DataGrid.tsx]: removed

File: src/datagrid/DataGrid.tsx (hunk 7 above)
Comment (akate1) [reply src/datagrid/DataGrid.tsx]: removed

File: src/datagrid/DataGrid.tsx (hunk 4 above)
Comment (akate1) [reply src/datagrid/DataGrid.tsx:1229]: done

File: src/datagrid/DataGrid.tsx (hunk 3 above)
Comment (akate1) [reply src/datagrid/DataGrid.tsx:1287]: done

**Review by girishdudhe — APPROVED** [review]
LGTM

**Pull request merged**